File: frontend/src/services/conversationService.ts
/**
 * Conversation Service
 * API client for conversation operations
 */

const API_URL = import.meta.env.VITE_API_URL || 'http://localhost:8000';

export interface ConversationFilters {
  client_id?: string;
  status?: 'active' | 'closed' | 'pending';
  priority?: 'Low' | 'Medium' | 'High';
  search?: string;
}

export interface ConversationCreate {
  client_id: string;
  status: 'active' | 'closed' | 'pending';
  channel: 'whatsapp' | 'email' | 'web';
  assigned_agent_id?: string;
  priority?: 'Low' | 'Medium' | 'High';
  summary?: string;
  tags?: string[];
}

export interface ConversationUpdate {
  status?: 'active' | 'closed' | 'pending';
  assigned_agent_id?: string;
  priority?: 'Low' | 'Medium' | 'High';
  summary?: string;
  tags?: string[];
}

export interface ConversationResponse {
  id: string;
  client_id: string;
  status: 'active' | 'closed' | 'pending';
  channel: 'whatsapp' | 'email' | 'web';
  assigned_agent_id: string | null;
  unread_count: number;
  priority: 'Low' | 'Medium' | 'High';
  start_date: string;
  last_update: string;
  summary: string | null;
  tags: string[];
  created_at: string;
  updated_at: string;
}

export interface ConversationListResponse {
  items: ConversationResponse[];
  total: number;
  page: number;
  limit: number;
  has_next: boolean;
}

class ConversationService {
  private getAuthToken(): string {
    const token = localStorage.getItem('token');
    if (!token) {
      throw new Error('No authentication token found');
    }
    return token;
  }

  private async request<T>(
    endpoint: string,
    options: RequestInit = {}
  ): Promise<T> {
    const token = this.getAuthToken();

    const response = await fetch(`${API_URL}${endpoint}`, {
      ...options,
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`,
        ...options.headers,
      },
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({ detail: 'Unknown error' }));
      throw new Error(error.detail || `HTTP ${response.status}`);
    }

    return response.json();
  }

  /**
   * Create new conversation
   */
  async createConversation(data: ConversationCreate): Promise<ConversationResponse> {
    return this.request<ConversationResponse>('/api/conversations', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  }

  /**
   * Get conversations with filters and pagination
   */
  async getConversations(
    page: number = 1,
    limit: number = 20,
    filters?: ConversationFilters
  ): Promise<ConversationListResponse> {
    const params = new URLSearchParams({
      page: page.toString(),
      limit: limit.toString(),
    });

    if (filters?.client_id) params.append('client_id', filters.client_id);
    if (filters?.status) params.append('status', filters.status);
    if (filters?.priority) params.append('priority', filters.priority);
    if (filters?.search) params.append('search', filters.search);

    return this.request<ConversationListResponse>(
      `/api/conversations?${params.toString()}`
    );
  }

  /**
   * Get conversation by ID
   */
  async getConversationById(id: string): Promise<ConversationResponse> {
    return this.request<ConversationResponse>(`/api/conversations/${id}`);
  }

  /**
   * Update conversation
   */
  async updateConversation(
    id: string,
    data: ConversationUpdate
  ): Promise<ConversationResponse> {
    return this.request<ConversationResponse>(`/api/conversations/${id}`, {
      method: 'PUT',
      body: JSON.stringify(data),
    });
  }

  /**
   * Update conversation status
   */
  async updateStatus(
    id: string,
    status: 'active' | 'closed' | 'pending'
  ): Promise<ConversationResponse> {
    return this.request<ConversationResponse>(
      `/api/conversations/${id}/status?status=${status}`,
      {
        method: 'PATCH',
      }
    );
  }

  /**
   * Delete conversation (soft delete)
   */
  async deleteConversation(id: string): Promise<void> {
    await this.request<void>(`/api/conversations/${id}`, {
      method: 'DELETE',
    });
  }

  /**
   * Mark conversation as read
   */
  async markAsRead(id: string): Promise<ConversationResponse> {
    return this.request<ConversationResponse>(
      `/api/conversations/${id}/mark-read`,
      {
        method: 'POST',
      }
    );
  }
}

// Export singleton instance
export const conversationService = new ConversationService();

File: backend/src/services/conversation_service.py
"""
Conversation service for business logic
"""
from typing import Optional, List
from uuid import UUID
from src.config.supabase import supabase_admin
from src.models.conversation import (
    ConversationCreate,
    ConversationUpdate,
    ConversationResponse
)
from src.utils.exceptions import NotFoundError, ValidationError
from src.utils.logger import logger


class ConversationService:
    """Service for conversation operations"""
    
    async def create_conversation(
        self,
        data: ConversationCreate
    ) -> ConversationResponse:
        """
        Create new conversation
        
        Args:
            data: Conversation data
            
        Returns:
            Created conversation
            
        Raises:
            ValidationError: Invalid data
        """
        try:
            # Prepare data
            conversation_data = data.model_dump()
            
            # Convert UUID to string for Supabase
            conversation_data['client_id'] = str(conversation_data['client_id'])
            if conversation_data.get('assigned_agent_id'):
                conversation_data['assigned_agent_id'] = str(conversation_data['assigned_agent_id'])
            
            # Insert into database
            response = supabase_admin.table("conversations").insert(
                conversation_data
            ).execute()
            
            if not response.data:
                raise ValidationError("Failed to create conversation")
            
            created = response.data[0]
            logger.info(f"Created conversation: {created['id']}")
            
            return ConversationResponse(**created)
            
        except Exception as e:
            logger.error(f"Error creating conversation: {str(e)}")
            raise ValidationError(f"Failed to create conversation: {str(e)}")
    
    async def list_conversations(
        self,
        page: int = 1,
        limit: int = 20,
        client_id: Optional[str] = None,
        status: Optional[str] = None,
        priority: Optional[str] = None,
        search: Optional[str] = None
    ) -> dict:
        """
        List conversations with filters and pagination
        
        Args:
            page: Page number
            limit: Items per page
            client_id: Filter by client
            status: Filter by status
            priority: Filter by priority
            search: Search in summary
            
        Returns:
            Paginated list of conversations
        """
        try:
            # Base query
            query = supabase_admin.table("conversations").select("*", count="exact")
            
            # Apply filters
            if client_id:
                query = query.eq("client_id", client_id)
            
            if status:
                query = query.eq("status", status)
            
            if priority:
                query = query.eq("priority", priority)
            
            if search:
                query = query.ilike("summary", f"%{search}%")
            
            # Pagination
            offset = (page - 1) * limit
            query = query.range(offset, offset + limit - 1)
            
            # Order by last_update (most recent first)
            query = query.order("last_update", desc=True)
            
            # Execute query
            response = query.execute()
            
            total = response.count or 0
            items = [ConversationResponse(**item) for item in response.data]
            has_next = total > (page * limit)
            
            logger.info(f"Listed {len(items)} conversations (page {page}/{limit})")
            
            return {
                "items": items,
                "total": total,
                "page": page,
                "limit": limit,
                "has_next": has_next
            }
            
        except Exception as e:
            logger.error(f"Error listing conversations: {str(e)}")
            raise
    
    async def get_conversation_by_id(
        self,
        conversation_id: str
    ) -> ConversationResponse:
        """
        Get conversation by ID
        
        Args:
            conversation_id: Conversation ID
            
        Returns:
            Conversation data
            
        Raises:
            NotFoundError: Conversation not found
        """
        try:
            response = supabase_admin.table("conversations").select("*").eq(
                "id", conversation_id
            ).single().execute()
            
            if not response.data:
                raise NotFoundError(f"Conversation {conversation_id} not found")
            
            logger.info(f"Retrieved conversation: {conversation_id}")
            
            return ConversationResponse(**response.data)
            
        except Exception as e:
            logger.error(f"Error retrieving conversation {conversation_id}: {str(e)}")
            if "not found" in str(e).lower() or "No rows found" in str(e):
                raise NotFoundError(f"Conversation {conversation_id} not found")
            raise
    
    async def update_status(
        self,
        conversation_id: str,
        status: str
    ) -> ConversationResponse:
        """
        Update conversation status
        
        Args:
            conversation_id: Conversation ID
            status: New status ('active', 'closed', 'pending')
            
        Returns:
            Updated conversation
            
        Raises:
            NotFoundError: Conversation not found
            ValidationError: Invalid status
        """
        try:
            # Validate status
            valid_statuses = ['active', 'closed', 'pending']
            if status not in valid_statuses:
                raise ValidationError(f"Invalid status. Must be one of: {valid_statuses}")
            
            # Verify conversation exists
            await self.get_conversation_by_id(conversation_id)
            
            # Update status
            response = supabase_admin.table("conversations").update({
                "status": status
            }).eq("id", conversation_id).execute()
            
            if not response.data:
                raise NotFoundError(f"Conversation {conversation_id} not found")
            
            updated = response.data[0]
            logger.info(f"Updated conversation status: {conversation_id} -> {status}")
            
            return ConversationResponse(**updated)
            
        except (NotFoundError, ValidationError):
            raise
        except Exception as e:
            logger.error(f"Error updating conversation status {conversation_id}: {str(e)}")
            raise ValidationError(f"Failed to update status: {str(e)}")
    
    async def update_conversation(
        self,
        conversation_id: str,
        data: ConversationUpdate
    ) -> ConversationResponse:
        """
        Update conversation
        
        Args:
            conversation_id: Conversation ID
            data: Data to update
            
        Returns:
            Updated conversation
            
        Raises:
            NotFoundError: Conversation not found
        """
        try:
            # Verify conversation exists
            await self.get_conversation_by_id(conversation_id)
            
            # Prepare data (only non-None fields)
            update_data = data.model_dump(exclude_unset=True)
            
            if not update_data:
                raise ValidationError("No data to update")
            
            # Convert UUIDs to strings
            if 'assigned_agent_id' in update_data and update_data['assigned_agent_id']:
                update_data['assigned_agent_id'] = str(update_data['assigned_agent_id'])
            
            # Update
            response = supabase_admin.table("conversations").update(
                update_data
            ).eq("id", conversation_id).execute()
            
            if not response.data:
                raise NotFoundError(f"Conversation {conversation_id} not found")
            
            updated = response.data[0]
            logger.info(f"Updated conversation: {conversation_id}")
            
            return ConversationResponse(**updated)
            
        except (NotFoundError, ValidationError):
            raise
        except Exception as e:
            logger.error(f"Error updating conversation {conversation_id}: {str(e)}")
            raise ValidationError(f"Failed to update conversation: {str(e)}")
    
    async def delete_conversation(
        self,
        conversation_id: str
    ) -> bool:
        """
        Delete conversation (soft delete by setting status to 'closed')
        
        Args:
            conversation_id: Conversation ID
            
        Returns:
            True if deleted successfully
            
        Raises:
            NotFoundError: Conversation not found
        """
        try:
            # Soft delete: set status to closed
            await self.update_status(conversation_id, 'closed')
            
            logger.info(f"Soft deleted conversation: {conversation_id}")
            
            return True
            
        except NotFoundError:
            raise
        except Exception as e:
            logger.error(f"Error deleting conversation {conversation_id}: {str(e)}")
            raise
    
    async def mark_as_read(
        self,
        conversation_id: str
    ) -> ConversationResponse:
        """
        Mark all messages in conversation as read (set unread_count to 0)
        
        Args:
            conversation_id: Conversation ID
            
        Returns:
            Updated conversation
            
        Raises:
            NotFoundError: Conversation not found
        """
        try:
            # Verify conversation exists
            await self.get_conversation_by_id(conversation_id)
            
            # Update unread_count to 0
            response = supabase_admin.table("conversations").update({
                "unread_count": 0
            }).eq("id", conversation_id).execute()
            
            if not response.data:
                raise NotFoundError(f"Conversation {conversation_id} not found")
            
            updated = response.data[0]
            logger.info(f"Marked conversation as read: {conversation_id}")
            
            return ConversationResponse(**updated)
            
        except NotFoundError:
            raise
        except Exception as e:
            logger.error(f"Error marking conversation as read {conversation_id}: {str(e)}")
            raise


# Global instance
conversation_service = ConversationService()
